compiler/vm/vm.nim

```diff
@@ -1156,86 +1448,182 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
-      let b = regs[rc].node
-      regs[ra].intVal = ord(containsSets(c.config, a, b) and not equalSets(c.config, a, b))
+      # a = b is proper subset of c
+      decodeBC(akInt)
+
+      var ret = true
+
+      # TODO: this is slow. Improve
+      for v in regs[rb].atomVal.setElems.items:
+        if v notin regs[rc].atomVal.setElems:
+          ret = false
+          break
+      
+
+      if ret: ret = regs[rb].atomVal.setElems.len != regs[rc].atomVal.setElems.len
+
+      regs[ra].intVal = ord(ret)
     of opcMulSet:
-      decodeBC(rkNode)
-      createSet(regs[ra])
-      move(regs[ra].node.sons,
-            nimsets.intersectSets(c.config, regs[rb].node, regs[rc].node).sons)
+      # a = intersection(b, c)
+      decodeBC(akSet)
+
+      # TODO: inefficient implementation. Improve
+
+      func addWhereBoth(dst: var seq[uint16]; a, b: openArray[uint16]) {.inline.} =
+        for v in a.items:
+          if v in b:
+            dst.add(v)
+
+      # Make sure `setElems` is empty
+      regs[ra].atomVal.setElems.setLen(0)
+
+      
+      var a = unsafeAddr regs[rb].atomVal.setElems
+      var b = unsafeAddr regs[rc].atomVal.setElems
+
+      # Make sure to loop over the smaller sequence in `addWhereBoth`
+      if b[].len < a[].len:
+          swap(a, b)
+
+      addWhereBoth(regs[ra].atomVal.setElems, a[], b[])
     of opcPlusSet:
-      decodeBC(rkNode)
-      createSet(regs[ra])
-      move(regs[ra].node.sons,
-           nimsets.unionSets(c.config, regs[rb].node, regs[rc].node).sons)
+      # a = union(b, c)
+      decodeBC(akSet)
+
+      # TODO: inefficient implementation. Improve
+
+      # Copy elements from b to tmp
+      var tmp = regs[rb].atomVal.setElems
+      tmp.add(regs[rc].atomVal.setElems)
+
+      regs[ra].atomVal.setElems = deduplicate(tmp)
     of opcMinusSet:
-      decodeBC(rkNode)
-      createSet(regs[ra])
-      move(regs[ra].node.sons,
-           nimsets.diffSets(c.config, regs[rb].node, regs[rc].node).sons)
+      # a = diff(b, c)
+      decodeBC(akSet)
+
+      # TODO: inefficient implementation. Improve
+
+      # Make sure `setElems` is empty
+      regs[ra].atomVal.setElems.setLen(0)
+
+      # Add all elements to output that are in b but not in c
+      for v in regs[rb].atomVal.setElems:
+        if v notin regs[rc].atomVal.setElems:
+          regs[ra].atomVal.setElems.add(v)
+
     of opcConcatStr:
-      decodeBC(rkNode)
-      createStr regs[ra]
-      regs[ra].node.strVal = getstr(regs[rb])
-      for i in rb+1..rb+rc-1:
-        regs[ra].node.strVal.add getstr(regs[i])
+      decodeBC(akString)
+      # Concat all values from register range rb..<rc together. Integer values are automatically treated as char
+
+      var tmp : array[1, char]
+
+      func appendstr(str: var string, r: TFullReg) =
+        case r.atomVal.kind
+        of akString:
+          str.add r.atomVal.strVal
+        of akInt:
+          # TODO: original code didn't check if int is in range 0..255. Should we do that now?
```

Comment:
I tend towards no.

There is a lot going on. First the byte code gen might have made a mistake. Second of this was just memory/CPU we'd have loaded the correct size into the register, as in we shouldn't have loaded 64 bits, but the 8 that are the char, then stored that. Locations are dynamic and all that.